src/models/event.js

```diff
@@ -327,17 +347,50 @@ utils.extend(module.exports.MatrixEvent.prototype, {
             );
         }
 
-        try {
-            crypto.decryptEvent(this);
-        } catch (e) {
-            console.warn(
-                `Error decrypting event (id=${this.getId()}): ${e}`,
+        if (this._decrypting) {
+            console.log(
+                `Event ${this.getId()} already being decrypted; queueing a retry`,
             );
+            this._retryDecryption = true;
+            return;
+        }
+
+        this._decrypting = true;
+
+        this._doDecryption(crypto).finally(() => {
+            this._decrypting = false;
+            this._retryDecryption = false;
```

Comment:
_retryDecryption is set when we have new data that might help fix a DecryptionError. It's never going to fix any other sort of error, so retrying the decryption will just cause more errors.